server: load stores in parallel during node startup

josh-hildred opened the pull request:
Previously, stores for initialized engines were loaded sequentially, resulting in long startup times. This commit adds the option to load stores for initialized engines in parallel. To disable parallel store start, set the environment variable `COCKROACH_CONCURRENT_STORE_START=false`.

Fixes: https://github.com/cockroachdb/cockroach/issues/114137

Release note (performance improvement): Allows stores to start in parallel reducing start times for nodes with many stores.

@erikgrinaker

josh-hildred committed:
server: load stores in parallel during node start

Previously, stores for initialized engines were loaded sequentially,
resulting in long startup times. This commit adds the option to load
stores for initialized engines in parallel. To disable parallel store
start, set the environment variable COCKROACH_CONCURRENT_STORE_START=false.

Fixes: #114137

Release note (performance improvement): Allows stores to start in parallel
reducing start times for nodes with many stores.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=115285) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=115285) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=115285) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/115285)
<!-- Reviewable:end -->

erikgrinaker commented on pkg/server/node.go:
```diff
@@ -651,17 +651,38 @@ func (n *Node) start(
+	// go routines will never wait.
+	c := make(chan error, len(state.initializedEngines))
+	for i := range state.initializedEngines {
+		engine := i
```
nit: consider pulling out `engine := state.initializedEngines[i]` here.

erikgrinaker commented on pkg/server/node.go:
```diff
@@ -651,17 +651,38 @@ func (n *Node) start(
+	// This also waits for all stores to finish starting.
+	for range state.initializedEngines {
+		if err := <-c; err != nil {
+			close(c)
```
We don't need to close here, we can and should leave that to GC. This will cause panics if a concurrent goroutine attempts to write to the channel after it's been closed. In general it's good practice to close on the sender side, never the receiver, but that's unnecessary here.

erikgrinaker replied:
There's still am unnecessary `close()` here on the receiver side. Let's not close the channel, we can leave it for GC.

erikgrinaker commented on pkg/server/node.go:
```diff
@@ -651,17 +651,38 @@ func (n *Node) start(
+	// Create stores in parallel from the engines that were already initialized. This uses a channel  to collect errors
+	// when starting stores. The channel is a buffered channel with the same size as the number of stores so worker
+	// go routines will never wait.
+	c := make(chan error, len(state.initializedEngines))
```
nit: since this remains in scope, let's give it a proper name, e.g. `engineErrC`.

erikgrinaker commented on pkg/server/node.go:
```diff
@@ -651,17 +651,38 @@ func (n *Node) start(
+	// Collect errors from the go routines and return the first error received.
+	// This also waits for all stores to finish starting.
+	for range state.initializedEngines {
+		if err := <-c; err != nil {
```
Let's use a `select` that also reacts to `ctx.Done()` and `n.stopper.ShouldQuiesce()`. Bare `<-` reads can lead to hangs if something goes wrong during shutdown.

erikgrinaker commented on pkg/server/node_test.go:
```diff
@@ -175,6 +175,64 @@ func TestBootstrapNewStore(t *testing.T) {
 	})
 }
 
+// TestStartManyStores starts a cluster with 20 stores and verifies all stores are stared correctly
```
```suggestion
// TestStartManyStores starts a cluster with 20 stores and verifies all stores are started correctly.
```

erikgrinaker reviewed (commented):
Thanks for the PR, this looks great!

I believe this should be safe, because we're already starting stores concurrently via `initializeAdditionalStores()`. Even so, let's add an escape hatch as an environment variable `COCKROACH_CONCURRENT_STORE_START=true` via `envutil.EnvOrDefaultBool()`, in case this ends up causing problems. An easy way to do that is to pass a 1-sized semaphore to `RunAsyncTaskEx` via `TaskOpts.Sem`, setting `WaitForSem: true`, passing a `nil` semaphore otherwise. We can add a TODO to remove the env var in 24.2, once we've gained some confidence.

Unfortunately, we don't have great multi-store test coverage, e.g. in end-to-end tests. That can be addressed separately.

Minor nits:

- Let's add a proper release note as "Release note (performance improvement): ...", which will be communicated to users.
- Commit titles should be lowercase and prefixed by the main package.
- The last commit should be squashed into the first.

Once that's addressed, I think we can mark this as ready and get this merged.

Thanks again!

josh-hildred commented:
Thank you for the review, @erikgrinaker!

I believe I have addressed all your comments, and it should be ready to go. However, please let me know if I misunderstood anything or if anything needs to be addressed further.

erikgrinaker commented on pkg/server/node.go:
```diff
@@ -651,17 +653,52 @@ func (n *Node) start(
+	// stores so worker go routines will never wait. By default, stores will be started concurrently. To start stores
+	// sequentially set the environment variable COCKROACH_CONCURRENT_STORE_START=false
+	var sem *quotapool.IntPool
+	var startStoresAsync = envutil.EnvOrDefaultBool("COCKROACH_CONCURRENT_STORE_START", true)
```
Let's move this up to a package global, below the "Cluster settings" block at the top of this file, along with the corresponding comment.

erikgrinaker commented on pkg/server/node.go:
```diff
@@ -651,17 +653,52 @@ func (n *Node) start(
-		s := kvserver.NewStore(ctx, n.storeCfg, e, &n.Descriptor)
-		if err := s.Start(workersCtx, n.stopper); err != nil {
-			return errors.Wrap(err, "failed to start store")
+	// Create stores from engines that are already initialized. This uses a channel to collect errors
+	// when starting stores in separate goroutines. The channel is a buffered channel with the same size as the number of
+	// stores so worker go routines will never wait. By default, stores will be started concurrently. To start stores
+	// sequentially set the environment variable COCKROACH_CONCURRENT_STORE_START=false
```
nit: break comments at 80 characters.

erikgrinaker reviewed (commented):
Looks good! Just a few nits and I think we can merge this.

(looks like VSCode prematurely submitted my review above for some reason)

erikgrinaker commented:
Looks good! Thanks for getting this across the line, and congratulations on landing your first PR! 🚀 

bors r+

erikgrinaker commented:
bors r+
